Add forward AD support for logsumexp, log_softmax, softmax, nll_loss, and cross_entropy

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73741

There are probably more perf improvements that can be made, for example reusing more quantities from forward, doing more things inplace, but in the spirit of improving coverage, this is probably OK for now.

Note: I didn't do anything with half_to_float, but CUDA (locally) hasn't complained yet

Differential Revision: [D34690141](https://our.internmc.facebook.com/intern/diff/D34690141)

soulitzer committed:
Add forward AD support for log_softmax, nll_loss, and cross_entropy

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for log_softmax, nll_loss, and cross_entropy"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for log_softmax, nll_loss, and cross_entropy"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for log_softmax, nll_loss, and cross_entropy"




[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for logsumexp, log_softmax, softmax, nll_loss, and cross_entropy"


There are probably more perf improvements that can be made, for example reusing more quantities from forward, doing more things inplace, but in the spirit of improving coverage, this is probably OK for now.

Note: I didn't do anything with half_to_float, but CUDA (locally) hasn't complained yet


[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -5200,6 +5200,19 @@ Tensor lu_factor_ex_jvp(
+  auto self_p_exp = self_p.exp();
+  auto sumexp_p = self_p_exp.sum(dim, keepdim);
+
+  if (areAnyTensorSubclassLike({self_p, self_t}) || self_t._is_zerotensor()) {
```
Are all the ops below supported by zero tensor efficiently?

soulitzer replied:
Not currently, according to fb.quip.com/Dv5hARK0xyTo, the missing ones are `sum` and `div`. 

Actually... we shouldn't need to worry about zero tensors for ops like this (that have only a single differentiable input) since we would be skipping the forward AD logic if the tensor does not have forward grad. The only worry though is if we have other formulas that reuse the jvp, for example softmax and logsoftmax here, but we are still OK for this PR because they also only have a single differentiable input.

albanD replied:
In that case we can remove the check?

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -5200,6 +5200,19 @@ Tensor lu_factor_ex_jvp(
   }
 }
 
+Tensor logsumexp_jvp(const Tensor& self_p, const Tensor& self_t, IntArrayRef dim, bool keepdim) {
```
I'm not sure if this is ok in terms of numerical stability.
I would expect this to have similar properties as the forward and I don't think the current implementation does.
Does the OpInfo have test with large values for the input?

soulitzer replied:
> Does the OpInfo have test with large values for the input?

Nope, but I can try to add.

> I'm not sure if this is ok in terms of numerical stability.

Yeah good catch, I'm trying to implement the exp normalize trick. It seems to work in python, but still investigating why the exact same code breaks in c++:
```
import torch

self_p = torch.rand(3, 4, 5)
self_t = torch.rand(3, 4, 5)

dim = (1, 2)
keepdim = False

def logsumexp_jvp_stable(self_p, self_t, dim, keepdim):
    self_p_exp = (self_p - torch.amax(self_p, dim, True)).exp()
    sumexp_p = self_p_exp.sum(dim, keepdim)

    return (self_p_exp * self_t).sum(dim, keepdim) / sumexp_p

def logsumexp_jvp(self_p, self_t, dim, keepdim):
    self_p_exp = self_p.exp()
    sumexp_p = self_p_exp.sum(dim, keepdim)

    return (self_p_exp * self_t).sum(dim, keepdim) / sumexp_p

print(torch.allclose(logsumexp_jvp(self_p, self_t, dim, keepdim), logsumexp_jvp_stable(self_p, self_t, dim, keepdim)))

class LogSumExp(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, dim, keepdim):
        ctx.dim = dim
        ctx.keepdim = keepdim
        ctx.save_for_forward(x)
        return torch.logsumexp(x, dim, keepdim)

    @staticmethod
    def jvp(ctx, x_t, _1, _2):
        dim = ctx.dim
        keepdim = ctx.keepdim
        x, = ctx.saved_tensors
        return logsumexp_jvp_stable(x, x_t, dim, keepdim)

a = torch.rand(2, 4, 5, 6, requires_grad=True, dtype=torch.double) * 1e6

print(torch.autograd.gradcheck(LogSumExp.apply, (a, (1, 2), True),
      check_forward_ad=True, check_backward_ad=False, check_batched_grad=False))
```

soulitzer committed:
Update on "Add forward AD support for logsumexp, log_softmax, softmax, nll_loss, and cross_entropy"


There are probably more perf improvements that can be made, for example reusing more quantities from forward, doing more things inplace, but in the spirit of improving coverage, this is probably OK for now.

Note: I didn't do anything with half_to_float, but CUDA (locally) hasn't complained yet


[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34690141).
